src/app/rpmostree-compose-builtin-tree.cxx

```diff
@@ -1430,3 +1441,164 @@ rpmostree_compose_builtin_tree (int             argc,
+                                      RpmOstreeCommandInvocation *invocation,
+                                      GCancellable   *cancellable,
+                                      GError        **error)
+{
+  g_autoptr(GOptionContext) context = g_option_context_new ("TREEFILE EXTYAML");
+  g_option_context_add_main_entries (context, common_option_entries, NULL);
+  g_option_context_add_main_entries (context, extensions_option_entries, NULL);
+
+  if (!rpmostree_option_context_parse (context,
+                                       NULL,
+                                       &argc, &argv,
+                                       invocation,
+                                       cancellable,
+                                       NULL, NULL, NULL, NULL, NULL,
+                                       error))
+    return FALSE;
+
+  if (argc < 3)
+    {
+      rpmostree_usage_error (context, "TREEFILE and EXTYAML must be specified", error);
+      return FALSE;
+    }
+  if (!opt_repo)
+    {
+      rpmostree_usage_error (context, "--repo must be specified", error);
+      return FALSE;
+    }
+  if (!opt_extensions_output_dir)
+    {
+      rpmostree_usage_error (context, "--output-dir must be specified", error);
+      return FALSE;
+    }
+
+  const char *treefile_path = argv[1];
+  const char *extensions_path = argv[2];
+
+  g_autofree char *basearch = rpm_ostree_get_basearch ();
+  g_autoptr(RORTreefile) treefile = ror_treefile_new (treefile_path, basearch, -1, error);
+  if (!treefile)
+    return glnx_prefix_error (error, "Failed to load treefile");
+
+  g_autoptr(OstreeRepo) repo = ostree_repo_open_at (AT_FDCWD, opt_repo, cancellable, error);
+  if (!repo)
+    return FALSE;
+
+  /* this is a similar construction to what's in rpm_ostree_compose_context_new() */
+  g_auto(GLnxTmpDir) cachedir_tmp = { 0, };
+  glnx_autofd int cachedir_dfd = -1;
+  if (opt_cachedir)
+    {
+      if (!glnx_opendirat (AT_FDCWD, opt_cachedir, TRUE, &cachedir_dfd, error))
+        return glnx_prefix_error (error, "Opening cachedir");
+    }
+  else
+    {
+      if (!glnx_mkdtempat (ostree_repo_get_dfd (repo),
+                           "tmp/rpm-ostree-compose.XXXXXX", 0700,
+                           &cachedir_tmp, error))
+        return FALSE;
+
+      cachedir_dfd = fcntl (cachedir_tmp.fd, F_DUPFD_CLOEXEC, 3);
+      if (cachedir_dfd < 0)
+        return glnx_throw_errno_prefix (error, "fcntl");
+    }
+
+  g_autofree char *base_rev = NULL;
+  if (!ostree_repo_resolve_rev (repo, opt_extensions_base_rev, FALSE, &base_rev, error))
+    return FALSE;
+
+  g_autoptr(GVariant) commit = NULL;
+  if (!ostree_repo_load_commit (repo, base_rev, &commit, NULL, error))
+    return FALSE;
+
+  g_autoptr(GPtrArray) packages =
+      rpm_ostree_db_query_all (repo, opt_extensions_base_rev, cancellable, error);
+  if (!packages)
+      return FALSE;
+
+  auto packages_mapping = std::make_unique<rust::Vec<rpmostreecxx::StringMapping>>();
```

Comment:
The string duping here is unfortunate (not that I care about performance for this much, but we should know how to do this better in other places).  Hmm...  Hard to solve without having a Rust API for `rpm_ostree_db_query_all()` which itself blocks on the Rust -> C++ calls.

Reply:
Yeah... it hurt a bit to write that.

One thing I tried doing was to hook up the bindings for `GPtrArray` -> `PtrArray` like you did for `GCancellable`, but I couldn't get it to work.